Webagent (#10)
airtop-ai/examples-python: examples/web_agent/README.md

# WEBAGENT

This is an agent that uses python to connect to Airtop, create sessions, execute prompts on the site and store the data in a local database.

## PREREQUISITES

- Python 3.10
- poetry==1.8.4
- poetry-core==1.9.1

> To install, just run the following after downloanding the code

Reviewer: typo:
```suggestion
To install, just run the following command after downloading the code
```

`poetry install`

These are the steps necessary to run the code:

> ## 1. CREATE WORKER

Reviewer: Remove the numbering on the titles. 1, 1.1 and 1.2 is a bit confusing

Author: Done

The worker.py is the core script of this agent.
It does the following in order:

> - It starts a session. It's necessary to supply your own API_KEY in the .env file;

Reviewer: Wrong name of the env var
```suggestion
- It starts a session. It's necessary to supply your own AIRTOP_API_KEY in the .env file;
```

Reviewer: We should also add `You can get one [here](https://portal.airtop.ai/api-keys) (sign-up required).`

Reviewer: > We should also add `You can get one [here](https://portal.airtop.ai/api-keys) (sign-up required).`

missed this change

Author: Done

> - It accesses the target site, defined in TARGET_URL in the .env file;

Reviewer: This instruction is incorrect. You are defining the TARGET_URL in the `worker.py` and not using the one on the .env file. We can keep the instruction but either move this variable to the example file or guide them to go to the `worker.py` file to edit. Also noticed you also define it on `utils.py`. We should define it in a single place.

Author: Done

> - It uses a prompt defined in the .env file to extract information from the site and requests login if needed;

Reviewer: This instruction is incorrect. You are defining the EXTRACT_DATA_PROMPT in the `worker.py` and not using the one on the .env file. We can keep the instruction but either move this variable to the example file or guide them to go to the `worker.py` file to edit. Also noticed you also define it on `utils.py`. We should define it in a single place.

Author: Done

- If it's the first time, the db and the worker are created and the result of running the prompt on the site is exhibited;
- If it's not the first time, then a comparison is made. The instruction for the comparison is on the file prompt_templates.py. If a comparison is made, it's inserted on the local db;
- The result of running the prompt on the site is inserted in the db;

## 1.1 FETCH RESULTS

In the file utils.py there are methods available to do the following operations:

> - describe the exhisting workers;

Reviewer: typo:
```suggestion
- describe the existing workers;
```

Author: Done

- delete a worker;

> - retireve the history of comparisons;

Reviewer: typo:
```suggestion
- retrieve the history of comparisons;
```

Author: Done

## 1.2 SCHEDULE WORKER

> It's possible to schedule a worker to run regularly. For that it's suggested to create a cron job or some other similar program that is able to wun worker.py with the desired recurrency.

Reviewer: ```suggestion
It's possible to schedule a worker to run regularly. For that, it's suggested to create a cron job or some other similar program that is able to run worker.py with the desired recurrence.
```

Reviewer: Are there instructions on how to schedule a worker that will exist in our docs that we can link to? @danielsht86 thoughts of if we should include it here

Author: Added the fix, waiting for instructions on where I can link it to

> 

Reviewer: Add a title:
`## Running the example`

Author: Done

Reviewer: I meant that you kept the `## SCHEDULE WORKER` section, but add `## Running the example` on line 39 above the text `To run the worker it's necessary run the command`

To run the worker it's necessary to do

> `python worker.py`

Reviewer: I see they could potentially run `python utils.py` to see what is stored in the DB though the results format is pretty hard to read.
![image](https://github.com/user-attachments/assets/5fc6e5bb-eb1a-4c84-a216-737b18096bff)

Did we intentionally want to give them the ability to see previous data in the DB?

If we intentionally want to keep this for them, we should format the results better and add this to the docs here:
```
## Looking at previous runs
To get details from the DB on all previous runs, run the following command:

`python utils.py`

Author: The idea is for them to be able to check the results. In this case it's using a json format that's transformed into a string to be stored. I must say that I don't know how to exhibit it better in the command line, so I'll add some code so the result is exported to an excel file

- poetry==1.8.4
- poetry-core==1.9.1

To install, just run the following command after downloading the code

`poetry install`

These are the steps necessary to run the code:

## CREATE WORKER

The worker.py is the core script of this agent.
It does the following in order:

> - It starts a session. It's necessary to supply your own AIRTOP_API_KEY in the worker.py script;

Reviewer: This should be in the .env file
```suggestion
- It starts a session. It's necessary to supply your own AIRTOP_API_KEY in the .env file;
```

- It accesses the target site, defined in TARGET_URL in the worker.py script;
- It uses a prompt defined in the worker.py script to extract information from the site and requests login if needed;
- If it's the first time, the db and the worker are created and the result of running the prompt on the site is exhibited;
- If it's not the first time, then a comparison is made. The instruction for the comparison is on the file prompt_templates.py. If a comparison is made, it's inserted on the local db;
- The result of running the prompt on the site is inserted in the db;

## FETCH RESULTS

In the file utils.py there are methods available to do the following operations:

- describe the existing workers;
- delete a worker;
- retrieve the history of comparisons;

## RUNNING THE EXAMPLE

It's possible to schedule a worker to run regularly. For that it's suggested to create a cron job or some other similar program that is able to run worker.py with the desired recurrence.

> To run the worker it's necessary to do

Reviewer: ```suggestion
To run the worker it's necessary run the command
```
